src/utils/utils_log.c

```diff
@@ -0,0 +1,222 @@
+                          "%Y-%m-%dT%H:%M:%S ", &tm_info);
+    }
+
+    if (loggerConfig.pid) {
+        ASSERT((intptr_t)sizeof(header) >
+               (h_pos - header)); //should never happen
+        h_pos += snprintf(h_pos, sizeof(header) - (h_pos - header),
+                          "PID:%-6lu TID:%-6lu ", (unsigned long)pid,
+                          (unsigned long)tid);
+    }
+
+    // We take twice header size here to ensure that
+    // we have space for log level and overflow string
+    // otherwise -Wformat-truncation might be thrown by compiler
+    char logLine[LOG_MAX + LOG_HEADER * 2];
+    snprintf(logLine, sizeof(logLine), "%s%-5s UMF] %s%s\n", header,
+             level_to_str(level), buffer, overflow);
+
+    fputs(logLine, loggerConfig.output);
+
+    if (level >= loggerConfig.flushLevel) {
+        fflush(loggerConfig.output);
+    }
+}
+
+#define MAX_FILE_PATH 256
+#define MAX_ENV_LEN 2048
+void utils_log_init(void) {
+    char envVar[MAX_ENV_LEN];
+
+    if (util_env_var("UMF_LOG", envVar, sizeof(envVar)) < 0) {
+        return;
+    }
+
+    if (match_env_var(envVar, "timestamp:yes", NULL)) {
+        loggerConfig.timestamp = 1;
+    } else if (match_env_var(envVar, "timestamp:no", NULL)) {
+        loggerConfig.timestamp = 0;
+    }
+
+    if (match_env_var(envVar, "pid:yes", NULL)) {
+        loggerConfig.pid = 1;
+    } else if (match_env_var(envVar, "pid:no", NULL)) {
+        loggerConfig.pid = 0;
+    }
+
+    if (match_env_var(envVar, "level:debug", NULL)) {
+        loggerConfig.level = LOG_DEBUG;
+    } else if (match_env_var(envVar, "level:info", NULL)) {
+        loggerConfig.level = LOG_INFO;
+    } else if (match_env_var(envVar, "level:warning", NULL)) {
+        loggerConfig.level = LOG_WARNING;
+    } else if (match_env_var(envVar, "level:error", NULL)) {
+        loggerConfig.level = LOG_ERROR;
+    }
+
+    if (match_env_var(envVar, "flush:debug", NULL)) {
+        loggerConfig.flushLevel = LOG_DEBUG;
+    } else if (match_env_var(envVar, "flush:info", NULL)) {
+        loggerConfig.flushLevel = LOG_INFO;
+    } else if (match_env_var(envVar, "flush:warning", NULL)) {
+        loggerConfig.flushLevel = LOG_WARNING;
+    } else if (match_env_var(envVar, "flush:error", NULL)) {
+        loggerConfig.flushLevel = LOG_ERROR;
+    }
+
+    const char *arg;
+    if (match_env_var(envVar, "output:stdout", NULL)) {
+        loggerConfig.output = stdout;
+    } else if (match_env_var(envVar, "output:stderr", NULL)) {
+        loggerConfig.output = stderr;
+    } else if (match_env_var(envVar, "output:file", &arg)) {
+        const char *argEnd = strstr(arg, ";");
+        char file[MAX_FILE_PATH + 1];
+        size_t len = 0;
+        if (argEnd) {
+            len = argEnd - arg;
+        } else {
+            len = strlen(arg);
+        }
+        if (len <= MAX_FILE_PATH) {
+            memcpy(file, arg, len);
+            file[len] = '\0';
+#ifdef _WIN32
```

Comment:
I saw similar code in ubench_fopen - as this is the second time we have to deal with fopen - could you move the ubench version to the utils and use it here?

Reply:
ubench.h is an external dependency with a different license - we should not refactor such code without serious reason.